File: src/economy.js
export function validateRank(rank, totalPlayers) {
  if (!Number.isInteger(rank) || rank < 1 || rank > totalPlayers) {
    throw new Error(`rank must be between 1 and ${totalPlayers}`);
  }
}

export function rankPointsFor(totalPlayers, rank) {
  validateRank(rank, totalPlayers);
  return totalPlayers - rank + 1;
}

export function goldPayoutFor(totalPlayers, rank) {
  validateRank(rank, totalPlayers);
  return totalPlayers - rank + 1;
}

export function applyRoundEconomy(players, roundRanks) {
  const totalPlayers = players.length;
  const rankByPlayer = new Map(roundRanks.map((r) => [r.playerId, r.rank]));

  for (const player of players) {
    const rank = rankByPlayer.get(player.id);
    if (!rank) {
      throw new Error(`Missing rank for player ${player.id}`);
    }

    const points = rankPointsFor(totalPlayers, rank);
    const gold = goldPayoutFor(totalPlayers, rank);

    player.totalPoints += points;
    player.gold += gold;
    player.lastRoundRank = rank;
    if (rank === 1) {
      player.firstPlaceCount += 1;
    }
  }
}

export function decideFinalWinner(players) {
  if (!players.length) {
    throw new Error('players cannot be empty');
  }

  const sorted = [...players].sort((a, b) => {
    if (b.totalPoints !== a.totalPoints) {
      return b.totalPoints - a.totalPoints;
    }
    if (b.firstPlaceCount !== a.firstPlaceCount) {
      return b.firstPlaceCount - a.firstPlaceCount;
    }
    if (a.lastRoundRank !== b.lastRoundRank) {
      return a.lastRoundRank - b.lastRoundRank;
    }
    return String(a.id).localeCompare(String(b.id));
  });

  return {
    winner: sorted[0],
    standings: sorted,
  };
}

File: Assets/Warlock/Scripts/Mvp/Economy.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Warlock.Mvp
{
    public sealed class PlayerProgress
    {
        public PlayerProgress(string id)
        {
            Id = id;
        }

        public string Id { get; }
        public int Gold { get; set; }
        public int TotalPoints { get; set; }
        public int FirstPlaceCount { get; set; }
        public int LastRoundRank { get; set; } = int.MaxValue;
        public HashSet<string> OwnedSkills { get; } = new();
        public Dictionary<string, string?> Loadout { get; } = LoadoutRules.CreateEmptyLoadout();
    }

    public readonly struct RoundRank
    {
        public RoundRank(string playerId, int rank)
        {
            PlayerId = playerId;
            Rank = rank;
        }

        public string PlayerId { get; }
        public int Rank { get; }
    }

    public readonly struct WinnerDecision
    {
        public WinnerDecision(PlayerProgress winner, IReadOnlyList<PlayerProgress> standings)
        {
            Winner = winner;
            Standings = standings;
        }

        public PlayerProgress Winner { get; }
        public IReadOnlyList<PlayerProgress> Standings { get; }
    }

    public static class EconomyRules
    {
        public static int RankPointsFor(int totalPlayers, int rank)
        {
            ValidateRank(totalPlayers, rank);
            return totalPlayers - rank + 1;
        }

        public static int GoldPayoutFor(int totalPlayers, int rank)
        {
            ValidateRank(totalPlayers, rank);
            return totalPlayers - rank + 1;
        }

        public static void ApplyRoundEconomy(IReadOnlyList<PlayerProgress> players, IReadOnlyList<RoundRank> roundRanks)
        {
            var rankByPlayer = roundRanks.ToDictionary(r => r.PlayerId, r => r.Rank);
            var totalPlayers = players.Count;

            foreach (var player in players)
            {
                if (!rankByPlayer.TryGetValue(player.Id, out var rank))
                {
                    throw new InvalidOperationException($"Missing rank for player {player.Id}");
                }

                var points = RankPointsFor(totalPlayers, rank);
                var gold = GoldPayoutFor(totalPlayers, rank);

                player.TotalPoints += points;
                player.Gold += gold;
                player.LastRoundRank = rank;

                if (rank == 1)
                {
                    player.FirstPlaceCount += 1;
                }
            }
        }

        public static WinnerDecision DecideFinalWinner(IReadOnlyList<PlayerProgress> players)
        {
            if (players.Count == 0)
            {
                throw new InvalidOperationException("players cannot be empty");
            }

            var standings = players
                .OrderByDescending(p => p.TotalPoints)
                .ThenByDescending(p => p.FirstPlaceCount)
                .ThenBy(p => p.LastRoundRank)
                .ThenBy(p => p.Id, StringComparer.Ordinal)
                .ToList();

            return new WinnerDecision(standings[0], standings);
        }

        private static void ValidateRank(int totalPlayers, int rank)
        {
            if (rank < 1 || rank > totalPlayers)
            {
                throw new ArgumentOutOfRangeException(nameof(rank), $"rank must be between 1 and {totalPlayers}");
            }
        }
    }
}
